### ISSUE TITLE
Support both internal and external gateways

### ISSUE THREAD

**Issue by szesch** [Kong/kong-operator#178]
## Problem Statement

For the cloud gateways project we are targeting supporting both north-south and east-west traffic for a gateway. In other words, a gateway may need to be both external and internal. This means we could have three possible configurations when configuring a gateway deployment:

1) External
2) Internal
3) Both

## Additional Information

This is to track discussions @sentinelleader has previously had with the gateway-operator team

## Acceptance Criteria
- [ ] A `DataPlane` can be deployed with either external or internal networking, or both

**Comment by shaneutt** [comment]
Based on https://github.com/Kong/gateway-operator/issues/802#issuecomment-1570887138 this should be one of our summit deliverables, so I'll drop them both in the relevant milestone.

@szesch when you say "gateway" in this context I just want to clarify that you're referring to the Kong Gateway via our `DataPlane` API, right? So for the purposes of your requirements, fields added to `DataPlane` that allow toggling between public and private LB are sufficient?

**Comment by szesch** [comment]
@shaneutt Yes, that's correct. This is for the `DataPlane`. As for the fields, we need to be able to create a `DataPlane` that  has public LB, private LB, or both (when a customer wants to use the same `Dataplane` for both north-south and east-west traffic).

**Comment by shaneutt** [comment]
> @shaneutt Yes, that's correct. This is for the `DataPlane`. As for the fields, we need to be able to create a `DataPlane` that has public LB, private LB, or both (when a customer wants to use the same `Dataplane` for both north-south and east-west traffic).

Sounds good, I've updated the AC a bit to match. Thank you :+1:

**Comment by shaneutt** [comment]
@szesch so as it turns out we apparently added the ability to modify the service type back in https://github.com/Kong/gateway-operator/pull/739. This unfortunately goes again the desire for opacity and to keep the `Service` an implementation detail (as we discussed recently in https://github.com/Kong/gateway-operator/issues/763 :confounded:) but given that it's already there does this in fact give you the tunable you need (e.g. `ClusterIP` for internal, `LoadBalancer` for both internal and external)?

**Comment by shaneutt** [comment]
Blocked waiting on clarifications, (re: https://github.com/Kong/gateway-operator/issues/803#issuecomment-1584948506)

**Comment by szesch** [comment]
Sorry, was on PTO. Being able to set the service to type `LoadBalancer` allows us to meet the requirement of creating an external **or** internal, but we still need to figure out how to handle the scenario when we need external **and** internal. For instance, imagine a customer wants to send their own customers traffic through Kong over the public internet but they also want to send traffic over a peered VPC or transit gateway for some internal use case. We need both an external and an internal load balancer for a single dataplane in that case.

**Comment by shaneutt** [comment]
> Sorry, was on PTO. Being able to set the service to type `LoadBalancer` allows us to meet the requirement of creating an external **or** internal, but we still need to figure out how to handle the scenario when we need external **and** internal. For instance, imagine a customer wants to send their own customers traffic through Kong over the public internet but they also want to send traffic over a peered VPC or transit gateway for some internal use case. We need both an external and an internal load balancer for a single dataplane in that case.

Thanks for the update. Using a `LoadBalancer` type `Service` will inherently provide both internal and external access, load-balanced according to the `Service` implementation (e.g. iptables, ipvs, or maybe ebpf if you use Cilium). You'll get a public `EXTERNAL-IP` as well as a `CLUSTER-IP`. Using the `DataPlane` status it should be possible to identify the public vs IP address and make a topology decision. Is this sufficient for your immediate needs or are there further requirements we need to accommodate for here?

**Comment by sentinelleader** [comment]
@shaneutt my understanding is that for LB service type, the clusterIP won't be from vpc cidr but from an internal ip range that's only been used cross node comms and cannot be accessed from a vpc peering. Incluster traffic works fine with clusterIP beacuse of the iptable magick and they are in same node network

https://github.com/aws/amazon-vpc-cni-k8s/issues/290#issuecomment-464266500 confirms my theory somewhat

**Comment by shaneutt** [comment]
Gotcha, this is the kind of use case stuff I was trying to dig for. So the requirement I'm digging for here is effectively:

- on AWS EKS I should be able to have both an internal and external load-balancers for a `DataPlane` simultaneously, and under this configuration **the internal one should be accessible via the VPC (e.g. so that peered VPCs across a multi-network environment can route to it)**

Sound right?

**Comment by szesch** [comment]
Yep!

**Comment by shaneutt** [comment]
Talked this over more with @sentinelleader and it sounds like what we may need to do is add the ability to deploy two `Services` such that both of them will be of type `LoadBalancer`, but one of them can be injected with the necessary annotations to tell AWS to make it an internal load-balancer which will only be routable via the VPC.

**Comment by shaneutt** [comment]
After discussing with @sentinelleader it sounds like this and it's sibling Kong/gateway-operator-archive#802 are not required for phase 0, but rather for API summit, so we'll move the milestones accordingly and that will allow us more time to consider our options.

**Comment by pmalek** [comment]
@sentinelleader @szesch 👋

Do you still require this for CG or have you managed to find another way to achieve this?

**Comment by sentinelleader** [comment]
@pmalek no this hasnt come up as a priority from any customers so far

**Comment by pmalek** [comment]
@sentinelleader Would this be fine with you to close this one and revisit when we get user demand for this?

**Comment by sentinelleader** [comment]
> [@sentinelleader](https://github.com/sentinelleader) Would this be fine with you to close this one and revisit when we get user demand for this?

yes 👍

**Comment by pmalek** [comment]
Closing, let's revisit when there's actual user demand for this.

**Comment by arm749** [comment]
@pmalek I don't have the use case for needing both an internal and external dataplane. However, I am trying to understand how to deploy an internal plane with an internal Azure load balancer. Similarly to what you can do with the annotation on the ingress controller service. I do not see any documentation that calls out how this can be done, nor is there a field in the Helm values to turn on an "internal" mode. Is this currently supported?

I did a workaround where I manually added the annotation to set it ```service.beta.kubernetes.io/azure-load-balancer-ipv4```

**Comment by pmalek** [comment]
@arm749 👋 

I'm not sure if this is what you meant but you can set ingress `Service` annotations with 

```
spec:
  network:
    services:
      ingress:
        annotations:
          foo: bar
```

https://github.com/Kong/kong-operator/blob/14c0c765605f29a41540cb0ce2c4b9bdc25ca189/config/samples/dataplane.yaml#L32-L36

**Comment by arm749** [comment]
> [@arm749](https://github.com/arm749) 👋
> 
> I'm not sure if this is what you meant but you can set ingress `Service` annotations with
> 
> ```
> spec:
>   network:
>     services:
>       ingress:
>         annotations:
>           foo: bar
> ```
> 
> [kong-operator/config/samples/dataplane.yaml](https://github.com/Kong/kong-operator/blob/14c0c765605f29a41540cb0ce2c4b9bdc25ca189/config/samples/dataplane.yaml#L32-L36)
> 
> Lines 32 to 36 in [14c0c76](/Kong/kong-operator/commit/14c0c765605f29a41540cb0ce2c4b9bdc25ca189)
> 
>  network: 
>    services: 
>      ingress: 
>        annotations: 
>          foo: bar

@pmalek I did not see that, thanks. I am using the operator, so it's the one provisioning the dataplane. But that lead me down the right path, and into the API docs for the GatewayConfiguration. I added the following to my GatewayConfiguration and got the desired effect. 😄 
```
 dataPlaneOptions:
    network:
      services:
        ingress:
          annotations:
            service.beta.kubernetes.io/azure-load-balancer-internal: "true"
